device/mellanox/x86_64-mlnx_lssn2700-r0/plugins/sfputil.py

```diff
@@ -149,3 +193,21 @@ def reset(self, port_num):
             return False
 
         return False
+
+    def get_transceiver_change_event(self, timeout=0):
```

Comment:
>get_transceiver_change_event [](start = 8, length = 28)

This is vendor independent code. Let's move it out of plugin.
Who is using it?

Reply:
@qiluo-msft this function is intended to let each vendor have their own implementation, because each vendor may have different ways to get the SFP plug in/out notification. Here is the implementation on mlnx platform.

Reply:
I would argue [https://github.com/Azure/SONiC/blob/gh-pages/doc/transceiver-monitor-hld.md](https://github.com/Azure/SONiC/blob/gh-pages/doc/transceiver-monitor-hld.md) is not a good design, at least for Mellanox platform.

Mellanox syncd docker could directly get SFP notification and write to STATE_DB. There is no benefit to write to redis once, blocking waiting on it in another container (pmon) and write redis again. The disadvantage is new dependencies (like in this file), fragile system, and hard to debug.

If this is true for Mellanox, how could it benefit other vendors?

---
In reply to: [200867307](https://github.com/Azure/sonic-buildimage/pull/1839#discussion_r200867307) [](ancestors = 200867307)

Reply:
I would agree with you if we only consider mlnx platform, and at the very beginning, my design is to post the transceiver info to DB from syncd since mlnx can get the sfp change event from there. But after discussing with Guohan and community,  we should have a common solution for all the vendors, design as common sfp change event API and left it to be implemented by each vendor.

Reply:
Totally understood. This is a perfect to time to revisit the design.


---
In reply to: [201230542](https://github.com/Azure/sonic-buildimage/pull/1839#discussion_r201230542) [](ancestors = 201230542)

Reply:
Just checking in -- are we or are we not revisiting the design? I believe the current design is the common solution for all vendors that @keboliu mentioned above. Therefore, I don't believe there is a need to revisit the design.

Reply:
@jleveque Yes, this the common solution for all vendors that have discussed.
@qiluo-msft would like to get your view on this.
more background information about this, for other vendors, may possible to get this event from sysfs, which will be very easy for them to implement this common API since sysfs is accessible from pmon. you may check with Hui who have some investigation on other platforms.